crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -0,0 +1,100 @@
+use crate::semantic_index::ast_ids::HasScopedAstId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::expression::Expression;
+use crate::semantic_index::symbol::{ScopeId, ScopedSymbolId, SymbolTable};
+use crate::semantic_index::symbol_table;
+use crate::types::{infer_expression_types, IntersectionBuilder, Type, TypeInference};
+use crate::Db;
+use ruff_python_ast as ast;
+use rustc_hash::FxHashMap;
+use std::sync::Arc;
+
+/// Return type constraint, if any, on `definition` applied by `test`.
+pub(crate) fn narrowing_constraint<'db>(
+    db: &'db dyn Db,
+    test: Expression<'db>,
+    definition: Definition<'db>,
+) -> Option<Type<'db>> {
+    all_narrowing_constraints(db, test)
+        .get(&definition.symbol(db))
+        .copied()
+}
+
+#[salsa::tracked]
+fn all_narrowing_constraints<'db>(
+    db: &'db dyn Db,
+    test: Expression<'db>,
+) -> NarrowingConstraints<'db> {
+    NarrowingConstraintsBuilder::new(db, test).finish()
+}
+
+type NarrowingConstraints<'db> = FxHashMap<ScopedSymbolId, Type<'db>>;
+
+struct NarrowingConstraintsBuilder<'db> {
+    db: &'db dyn Db,
+    expression: Expression<'db>,
+    constraints: NarrowingConstraints<'db>,
+}
+
+impl<'db> NarrowingConstraintsBuilder<'db> {
+    fn new(db: &'db dyn Db, expression: Expression<'db>) -> Self {
+        Self {
+            db,
+            expression,
+            constraints: NarrowingConstraints::default(),
+        }
+    }
+
+    fn finish(mut self) -> NarrowingConstraints<'db> {
+        if let ast::Expr::Compare(expr_compare) = self.expression.node(self.db).node() {
+            self.add_expr_compare(expr_compare);
+        }
+        // TODO other test expression kinds
+
+        self.constraints.shrink_to_fit();
+        self.constraints
+    }
+
+    fn symbols(&self) -> Arc<SymbolTable> {
+        symbol_table(self.db, self.scope())
+    }
+
+    fn scope(&self) -> ScopeId<'db> {
+        self.expression.scope(self.db)
+    }
+
+    fn inference(&self) -> &TypeInference<'db> {
+        infer_expression_types(self.db, self.expression)
+    }
```

Comment:
Are these operations we expect to call often? Would it make sense to query them once instead of going through salsa multiple times?

Reply:
It would be really nice if we could think of calling a cached Salsa query as "free", but I guess we can't :/ Yeah, I'll cache these on the builder.

Reply:
On looking at this more, for a given `NarrowingConstraintsBuilder` we'd expect to usually call any of these zero or one times; if the constraint is particularly complex, maybe several times. I moved the calls outside the loop over comparators as you suggested below, but I don't think further caching of these is worth it. We could try it later once we have more kinds of constraints supported.

Unconditionally querying these on construction of a `NarrowingConstraintsBuilder` is like a 10-15% regression, because currently most of the time we query these zero times (because the constraint is not in the form `x is not y`, which is the only constraint form we support in this PR.)